homeassistant/components/modbus/modbus.py

```diff
@@ -189,41 +212,56 @@ def connect(self):
     def read_coils(self, unit, address, count):
         """Read coils."""
         with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            return self._client.read_coils(address, count, **kwargs)
+            return _read_cached(
+                self._client, "coils", unit, address, count, int(time.time())
+            )
 
     def read_discrete_inputs(self, unit, address, count):
         """Read discrete inputs."""
         with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            return self._client.read_discrete_inputs(address, count, **kwargs)
+            return _read_cached(
+                self._client, "discrete_inputs", unit, address, count, int(time.time())
+            )
 
     def read_input_registers(self, unit, address, count):
         """Read input registers."""
         with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            return self._client.read_input_registers(address, count, **kwargs)
+            return _read_cached(
+                self._client, "input_registers", unit, address, count, int(time.time())
+            )
 
     def read_holding_registers(self, unit, address, count):
         """Read holding registers."""
         with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            return self._client.read_holding_registers(address, count, **kwargs)
+            return _read_cached(
+                self._client,
+                "holding_registers",
+                unit,
+                address,
+                count,
+                int(time.time()),
+            )
 
     def write_coil(self, unit, address, value):
         """Write coil."""
         with self._lock:
             kwargs = {"unit": unit} if unit else {}
             self._client.write_coil(address, value, **kwargs)
+            # invalidate cache on write
+            _read_cached.cache_clear()
 
     def write_register(self, unit, address, value):
         """Write register."""
         with self._lock:
             kwargs = {"unit": unit} if unit else {}
             self._client.write_register(address, value, **kwargs)
+            # invalidate cache on write
+            _read_cached.cache_clear()
 
     def write_registers(self, unit, address, values):
         """Write registers."""
         with self._lock:
             kwargs = {"unit": unit} if unit else {}
             self._client.write_registers(address, values, **kwargs)
+            # invalidate cache on write
+            _read_cached.cache_clear()
```

Comment:
again.

In general I think all your changes to this file apart from the 2 loops either needs a lot more explanation and test, or be removed.